### ISSUE TITLE
Incorrect widget layout rounding

### ISSUE THREAD

**Issue by stfx** [ocornut/imgui#1858]
Current (at certain resolutions):
![swarmlake_2018_06_04_17_57_26_107](https://user-images.githubusercontent.com/341158/40928405-9a773068-6821-11e8-9959-2abf4a4e77b7.png)

Expected:
![swarmlake_2018_06_04_18_15_34_765](https://user-images.githubusercontent.com/341158/40928935-529238cc-6823-11e8-8f7c-eb2e90e0dc29.png)

Reproduce:
```c++
float baseScale = 720.f;
float uiScale = glm::min(ImGui::GetIO().DisplaySize.x, ImGui::GetIO().DisplaySize.y) / baseScale;
ImGui::GetStyle().Colors[ImGuiCol_Button] = ImVec4(0.f, 0.f, 0.f, 0.65f);
ImGui::GetStyle().WindowPadding = ImVec2(2.f, 2.f) * uiScale;
ImGui::GetStyle().FramePadding = ImVec2(32.f, 10.f) * uiScale;
ImGui::GetStyle().ItemSpacing = ImVec2(0.f, 0.f) * uiScale;
ImGui::Button(...)
ImGui::Button(...)
...
```

Details:
Widgets are overlapping by 1px (horizontal black lines in screenshot) or having a gap of 1px (vertically between back and reset button in screenshot) due to incorrect rounding which is noticeable at zero ItemSpacing at certain resolutions when scaling the UI.

Version: Dear ImGui 1.61 release (and earlier like 1.50 so not a regression)

**Comment by stfx** [comment]
I tried using `ImFloor` like in `ScaleAllSizes()` in the past and just now and confirmed that all style size values are rounded values however I am still getting these issues which still makes me believe that there are some rounding issues somewhere:

![swarmlake_2018_06_04_18_47_05_403](https://user-images.githubusercontent.com/341158/40930371-d297fcce-6827-11e8-9253-8f5b97291110.png)

Style | Value| Type
------|-------|-------
Alpha | 1.00000000 | float
WindowPadding | {x=1.00000000 y=1.00000000 } | ImVec2
WindowRounding | 0.000000000 | float
WindowBorderSize | 0.000000000 | float
WindowMinSize | {x=0.000000000 y=0.000000000 } | ImVec2
WindowTitleAlign | {x=0.000000000 y=0.500000000 } | ImVec2
ChildRounding | 0.000000000 | float
ChildBorderSize | 1.00000000 | float
PopupRounding | 0.000000000 | float
PopupBorderSize | 1.00000000 | float
FramePadding | {x=13.0000000 y=13.0000000 } | ImVec2
FrameRounding | 0.000000000 | float
FrameBorderSize | 0.000000000 | float
ItemSpacing | {x=13.0000000 y=13.0000000 } | ImVec2
ItemInnerSpacing | {x=13.0000000 y=13.0000000 } | ImVec2
TouchExtraPadding | {x=0.000000000 y=0.000000000 } | ImVec2
IndentSpacing | 21.0000000 | float
ColumnsMinSpacing | 6.00000000 | float
ScrollbarSize | 16.0000000 | float
ScrollbarRounding | 0.000000000 | float
GrabMinSize | 15.0000000 | float
GrabRounding | 0.000000000 | float
ButtonTextAlign | {x=0.500000000 y=0.500000000 } | ImVec2
DisplayWindowPadding | {x=20.0000000 y=20.0000000 } | ImVec2
DisplaySafeAreaPadding | {x=3.00000000 y=3.00000000 } | ImVec2
MouseCursorScale | 1.00000000 | float

**Comment by stfx** [comment]
I tried using `ImFloor` on style sizes values as well as all widget positions and sizes in the past and now again however the same issue is still happening (on slightly different resolutions) which makes me believe that there are rounding issues within Dear ImGui widget rendering

**Comment by stfx** [comment]
You are right, I forgot using `ImFloor` when loading fonts as well. However I am scaling down the window font scale for the shown settings window like so `ImGui::SetWindowFontScale(0.75f);` which however causes those issues and I am not sure how to best fix this since this value is directly accessed by the widgets in `ImGui::CalcTextSize`. Any ideas?

**Comment by ocornut** [comment]
ImGuiWindow::CalcFontSize doesn't round so it is probably indeed going to be a problem. I'll probably be running into the problem more often with work on DPI in the Viewport branch, as the size per-window should technically be based on the DpiScale for that viewport/monitor.

The workaround may be to round the return value of ImGuiWindow:CalcFontSize but I'll probably wait until I have a more serious test bed for DPI scaling.

TL;DR; smooth, continuous and perfectly-linear scaling of UI will probably never work as long as we round anything, but it's probably not a big issue.
